### ISSUE TITLE
Clean up complex number arithmetic and use complex.h

### ISSUE THREAD

**Issue by billyauhk** [OSGeo/PROJ#2437]
**TLDR: Should we adopt complex.h and clean up the code?**

I was looking if I could add more projections mentioned in https://www.cs.dartmouth.edu/~doug/wallpaper.pdf into the library. A number of them involve elliptic functions with complex arguments or output. I now have self-sufficient C file, only depend on the "double complex" type from <complex.h>.

Complex number arithmetic in the existing code base is sometimes implemented with `struct COMPLEX`, sometimes the author splits the two components on their own (such as in `clenS()` in `tmerc.cpp`).

**RFC**: Since complex number arithmetic is now supported in C99/C11, **should I align all existing code to use the complex.h (and the "double complex" type) first** before adding new projections which will extensively use complex numbers?

P.S.
1. Somehow the projections in the paper was implemented https://web.archive.org/web/20110606175925/https://www.cs.dartmouth.edu/~doug/map.tar.gz (licensing terms not clear). Thus my self-sufficient code avoided them in order to maintain license compatibility.
2. These projections belong to a family which contains all 5 projections in `adams.cpp`.

**Comment by kbevers** [comment]
> RFC: Since complex number arithmetic is now supported in C99/C11, should I align all existing code to use the complex.h (and the "double complex" type) first before adding new projections which will extensively use complex numbers?

This sounds like a good opportunity to harmonize and clean the code base, so I'd say yes. I am interested to hear what @rouault and @cffk thinks though.

**Comment by rouault** [comment]
I'm not familiar with the use of complex data type from complex.h, but it seems that Microsoft might not be fully standard : https://docs.microsoft.com/en-us/cpp/c-runtime-library/complex-math-support?view=msvc-140 ( """The compiler does not directly support a complex or _Complex keyword, therefore the Microsoft implementation uses structure types to represent complex numbers."""). Not completely sure if they are difference between C and C++ modes, but our source code is C++ now. I'd suggest pragmatism to @billyauhk and issue a dummy pull requet hacking an existing file to just exercise the anticipated use of complex number and see how CI reacts

**Comment by cffk** [comment]
I only know about the C++ complex type and this seems to be solid enough.  It would be a good idea to use this as the container for complex number instead of some home-brewed struct.  There will be no cost to using `real(z)` and `imag(z)` instead of `z.r` and `z.i`.  Presumably the complex implementations of `*`, `+`, and `-`, are just as good as open-coded versions.  I'm not sure about `/` though; there may be times when this has to be done carefully.  In my experience (from 2017), computing `sin(z)` and `cos(z)` together doesn't do as good a job as computing these in terms of `sin/cos + sinh/cosh` of the real quantities.
I has scared by Microsoft's statement (in the link that @rouault posted): "These functions are implemented to balance performance with correctness."  I regard this as something of a red flag.  I hope MS's implementation of complex numbers in C++ doesn't take shortcuts.

**Comment by billyauhk** [comment]
Thanks for the comments. Have searched for complex arithmetic at HEAD and apparently I can hack some small files first as suggested by @rouault.
